proposals/NNNN-task-priority-escalation-apis.md

```diff
@@ -0,0 +1,237 @@
+
+## Motivation
+
+Generally developers can and should rely on the automatic task priority escalation happening transparently–at least for as long as all tasks necessary to escalate are created using structured concurrency primitives (task groups and `async let`). However, sometimes it is not possible to entirely avoid creating an unstructured task. 
+
+One such example is the async sequence [`merge`](https://github.com/apple/swift-async-algorithms/blob/main/Sources/AsyncAlgorithms/AsyncAlgorithms.docc/Guides/Merge.md) operation from the [swift-async-algorithms](https://github.com/apple/swift-async-algorithms/) project where the implementation is forced to create an unstructured task for iterating the upstream sequences, which must outlive downstream calls. These libraries would like to participate in task priority escalation to boost the priority of the upstream consuming task, however today lack the API to do so.
+
+```swift
+// SIMPLIFIED EXAMPLE CODE
+// Complete source: https://github.com/apple/swift-async-algorithms/blob/main/Sources/AsyncAlgorithms/Merge/MergeStorage.swift
+
+struct AsyncMergeSequenceIterator: AsyncIterator {
+  struct State {
+    var task: Task<Void, any Error>? // unstructured upstream consumer task
+    var buffer: Deque<Element>
+    var upstreamContinuations: [UnsafeContinuation<Void, Error>]
+    var downstreamContinuation: UnsafeContinuation<Element?, Error>?
+  }
+  
+  let state = Mutex<State>(State())
+  
+  func next() async throws {
+    self.state.withLock { state in
+      if state.task == nil {
+        state.task = Task {
+          // Consume from the base iterators
+          // ... 
+        }
+      }
+    }
+
+    if let element = self.state.withLock { $0.buffer.popFirst() } {
+      return element
+    } else {
+      // We are handling cancellation here and need to handle task escalation here as well
+      try await withTaskCancellationHandler { 
+        // HERE: need to handle priority escalation and boost `state.task`
+        try await withCheckedContinuation { cont in
+          self.state.withLock { $0.consumerContinuation = cont }
+        }
+      } onCancel: {
+        // trigger cancellation of tasks and fail continuations
+      }
+    }
+  }
+}
+```
+
+The above example showcases a common pattern: often an unchecked continuation is paired with a Task used to complete it. Around the suspension on the continuation, waiting for it to be resumed, developers often install a task cancellation handler in order to potentially break out of potentially unbounded waiting for a continuation to be resumed. Around the same suspension (marked with `HERE` in the snippet above), we might want to insert a task priority escalation handler in order to priority boost the task that is used to resume the continuation. This can be important for correctness and performance of such operations, so we should find a way to offer these libraries to participate in task priority handling.
```

Comment:
nit:

> often an unchecked continuation

there's a mix of checked & unchecked continuations in the above example. maybe drop the additional descriptor and just go with 'continuation'.

> so we should find a way to offer these libraries [a mechanism] to participate in task priority handling.